### ISSUE TITLE
SMILES issue (Sulphur 7 valence)

### ISSUE THREAD

**Issue by vedina** [cdk/cdk#279]
We have an issue reported that RdKit gives error, when parsing this SMILES 

```
BrC1=CC=C([SH](O)(=O)=CC=2OC(C(=O)N3CCOCC3)=CC2)C=C1
# Explicit valence for atom # 5 S, 7, is greater than permitted

```
 but CDK (via AMBIT )  parses it [fine (along with others) ](https://apps.ideaconsult.net/excape/depict?search=BrC1%3DCC%3DC%28%5BSH%5D%28O%29%28%3DO%29%3DCCC%28%3DO%29N2CCN%28CC2%29C3%3DCC%3DCC%3DC3%29C%3DC1)

Even the atom types are assigned. 

````
		String smi = "BrC1=CC=C([SH](O)(=O)=CC=2OC(C(=O)N3CCOCC3)=CC2)C=C1";
		SmilesParser parser = new SmilesParser(
				SilentChemObjectBuilder.getInstance());
	
		IAtomContainer mol = parser.parseSmiles(smi);
		AtomContainerManipulator.percieveAtomTypesAndConfigureAtoms(mol);
		for (IAtom a : mol.atoms()) {
			System.out.println(a.getAtomTypeName());
		}
````

The S atom type is `S.sp3d1`. 

Do you consider the SMILES is valid ?

**Comment by johnmay** [comment]
The valence is invalid but the SMILES is fine. This should be a standardisation failure but **not** fail but not a SMILES error.  RDKit is fussy - sometimes incorrectly ([PF6](https://en.wikipedia.org/wiki/Hexafluorophosphate)) was rejected for a long time but this time correctly.  RDKit has a concept of sanitisation which is applied by default on molecule input. Each to their own, personally I believe checks should be done after a valid molecule* is handed back to the user so you can choose to fix them.

Perhaps the atom typing should be undefined... but I wouldn't generally rely on that for validating/correcting molecules. I thought AMBIT had it's own standardisation? I remember Nikolay mentioning this at the MIOSS meeting last year.

*Kekulization complicates this as you have undefined bond order which may/may not be okay. In this case everything is define on input.

**Comment by johnmay** [comment]
...corrected some double negatives. Sorry doing too many things at once.

**Comment by vedina** [comment]
Yes it [does have own standardisation](http://ambit.sourceforge.net/ambitcli_standardisation.html). We probably are missing something in this case (I assumed the atom type will be undefined).

**Comment by johnmay** [comment]
**S.sp3d1** should be six valent. Looks like the implicit hydrogen is not being considered.

```java
<at:AtomType rdf:ID="S.sp3d1">
    <at:formalCharge>0</at:formalCharge>
    <at:hasElement rdf:resource="&elem;S"/>
    <at:lonePairCount>0</at:lonePairCount>
    <at:hybridization rdf:resource="&at;sp3d1"/>
    <at:formalBondType rdf:resource="&bo;single"/>
    <at:formalBondType rdf:resource="&bo;single"/>
    <at:formalBondType rdf:resource="&bo;single"/>
    <at:formalBondType rdf:resource="&bo;single"/>
    <at:formalBondType rdf:resource="&bo;double"/>
   </at:AtomType>  
```

**Comment by vedina** [comment]
Hmm, indeed.  Have to look at the atom typing code then.

**Comment by egonw** [comment]
I'll have a look at the atom typer code... here's some graphical info:

http://cdkdepict-openchem.rhcloud.com/depict/bow/svg?smi=BrC1=CC=C([SH](O)(=O)=CC=2OC(C(=O)N3CCOCC3)=CC2)C=C1&abbr=off&suppressh=true&showtitle=false&sma=c-%5BSH%5D(O)(%3DO)%3DC&zoom=1.3&annotate=none

**Comment by johnmay** [comment]
Without the erroneous hydrogen it should be S.sp3.4. We were actually discussing similar cases last week at NextMove. In a standardisation filter 5 valent nitrogens with a hydrogen should probably be rejected (``CN(C)=C``) as normally it's a charge missed. Likewise 4/6 valent sulphur with a H are normally wrong (related bug report from Daniel to Indigo: https://github.com/epam/Indigo/issues/67).

**Comment by egonw** [comment]
I like to keep it open... if the input SMILES is in error, then the perception code should return a "null"... more specifically, the code does a final check to see if the number of bound atoms matches... but all in all, this final match does not notice the type is incorrectly perceived... I think I need to fix my code to return a null...

**Comment by johnmay** [comment]
Sorry... click close and comment by mistake. My bad

**Comment by johnmay** [comment]
May affects tests but valence was not correctly calculated: https://github.com/cdk/cdk/pull/280

**Comment by ntk73** [comment]
I discovered that the problem in our case is caused by function: CDKHydrogenAdde.addImplicitHydrogens(). It turns out that for the cases like this  *-S(=O)(=C)-*, an extra H atom is added where if two double bonds are connected to oxygen atoms (e.g.  *-S(=O)(=О)-*) function seems to be working OK.

**Comment by johnmay** [comment]
You should rarely need to call addImplicitHydrogens, definitely do not call it for reading in molecules from SMILES/Molfile. XYZ it is needed, CML sometimes.

**Comment by vedina** [comment]
@johnmay it's not called after reading from a file, but after structure changes (e.g. applying SMIRKS, tautomerisation)

**Comment by johnmay** [comment]
Yes you shouldn't call it for that either - addImplicitHydrogens can change the MF you probably never want this. Your SMIRKS should account for the valence - particularly if you're moving hydrogens around. SMIRKS can be tricky to get right, See Slide 22 here https://www.slideshare.net/NextMoveSoftware/rdkit-six-notsoeasy-pieces-rdkit-ugm-2016 - The absence of the charge on the right does not mean set charge to 0. You must explicitly specify this.

**Comment by vedina** [comment]
Yep, Nick added an option to to switch on/of using the `addImplicitHydrogens` when generating tautomers, and it is indeed not really needed.  So far there is no issue with the predefined [SMIRKS](https://svn.code.sf.net/p/ambit/code/trunk/ambit2-all/ambit2-smarts/src/main/resources/ambit2/smirks/smirks.json).

Still it will be nice if `addImplicitHydrogens` works regardless of if where it is being called :)

**Comment by johnmay** [comment]
You should bound your valences:

```
[*:1][N:2](=[O:3])=[O:4]>>[*:1][N+:2](=[O:3])[O-:4]
```

That will also match hypervalent nitros, betst to bound the left with D/h or X/H.

```
[ND3h0:2](=[OD1h0:3])=[OD1h0:4]>>[N+:2](=[O:3])-[O-:4]
```

**Comment by vedina** [comment]
Thanks!

**Comment by johnmay** [comment]
> Still it will be nice if addImplicitHydrogens works regardless of if where it is being called :)

It's a misnomer unfortunately. A more verbose name ``deleteAndReAddHydrogensToNormalValence()`` would not be nice. A lot of the confusion comes from the *implicit* usage. There are implicit hydrogens and implicit implicit hydrogens. CDK `getImplicitHydrogenCount()` hydrogens aren't implicit they are *labelled* or *contracted* (singly implicit). What you really want a method to do (and is implementable on top of what CDK provides) as add hydrogens when none have been defined right? (Doubly implicit). There are lots of conversion bugs in toolkits because of this mixup.

**Comment by johnmay** [comment]
Now resolved.

**Comment by vedina** [comment]
For the record:  a similar SMILES with extra hydrogen `[SH]` is obtained when reading [this PubChem SDF](https://pubchem.ncbi.nlm.nih.gov/compound/102023954) and applying `addImplicitHydrogens`  (with CDK 1.5.13) .  Most probably reproducible with newer releases as [CDKAtomTypeMatcher#perceiveSulhur](https://github.com/cdk/cdk/blob/162bce94974332214791f118f02c4c8586fa1a4a/base/core/src/main/java/org/openscience/cdk/atomtype/CDKAtomTypeMatcher.java#L1282) has not changed.

**Comment by johnmay** [comment]
Don't call addImplicitHydrogens :-). Hopefully writing good SMIRKS handling this Xmas.

**Comment by vedina** [comment]
yep - how does one add hydrogens then :)

**Comment by johnmay** [comment]
```java
String url = "https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/102023954/record/SDF/?record_type=2d&response_type=display";
MDLV2000Reader reader   = new MDLV2000Reader(new URL(url).openStream());
AtomContainer  molecule = reader.read(new AtomContainer());
System.out.println(new SmilesGenerator().create(molecule));
```
Result:
```
S([O])(=O)(C(C([H])([H])[H])(C([H])([H])[H])C([H])([H])[H])=C(C#N)[H].[Li] |^1:1,20|
```
![](http://simolecule.com/cdkdepict/depict/bow/svg?smi=S([O])(=O)(C(C([H])([H])[H])(C([H])([H])[H])C([H])([H])[H])=C(C%23N)[H].[Li]%20%7C%5E1%3A1%2C20%7C&w=80&h=50&abbr=on&hdisp=provided&showtitle=false&zoom=1.6&annotate=none)

Yes if you perceive atom types and add explicit hydrogens it gets it wrong but as shown above you don't need to do that! The hydrogens are already there.

```java
String url = "https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/102023954/record/SDF/?record_type=2d&response_type=display";
        MDLV2000Reader reader   = new MDLV2000Reader(new URL(url).openStream());
        AtomContainer  molecule = reader.read(new AtomContainer());
        AtomContainerManipulator.percieveAtomTypesAndConfigureAtoms(molecule); // no
        CDKHydrogenAdder.getInstance(SilentChemObjectBuilder.getInstance()).addImplicitHydrogens(molecule); // no!
        AtomContainerManipulator.convertImplicitToExplicitHydrogens(molecule);
        System.out.println(new SmilesGenerator().create(molecule));
```

![](http://simolecule.com/cdkdepict/depict/bow/svg?smi=S([O])(=O)(C(C([H])([H])[H])(C([H])([H])[H])C([H])([H])[H])(=C(C%23N)[H])[H].[Li]%20%7C%5E1%3A1%2C21%7C&w=80&h=50&abbr=on&hdisp=provided&showtitle=false&zoom=1.6&annotate=none)

**Comment by johnmay** [comment]
Missed out ```AtomContainerManipulator.convertImplicitToExplicitHydrogens(molecule);``` from the first snippet but updated now.

**Comment by vedina** [comment]
@johnmay , I know hydrogens are there in this file, but in a generic workflow processing millions of such files you don't know if all H are there, none at all, or only some of them.

**Comment by johnmay** [comment]
No, they are always there for valid MDL files.

**Comment by vedina** [comment]
that's exactly the problem, I am talking about potentially-invalid MDL files found on the wild ... seen tons of these

**Comment by johnmay** [comment]
Fixed via #399... but I emphasise again your workflow should not depend on the CDK atom type handling.

**Comment by vedina** [comment]
CDK atom typing is a powerful thing,  I disagree it has to be ignored :)
